drivers/node/ssh/ssh.go

```diff
@@ -154,11 +155,10 @@ func (s *SSH) GetDeviceMapperCount(n node.Node, timerange time.Duration) (int, e
 	return count, nil
 }
 
-// Init initializes SSH node driver
-func (s *SSH) Init(nodeOpts node.InitOptions) error {
+func (s *SSH) updateDriver() error {
 	var err error
-	s.specDir = nodeOpts.SpecDir
 
+	// The namespace of the postworx-service in the current context
```

Comment:
not: it should be `portworx-service`

Reply:
I dont see this change. Please do not resolve conversation until you have fixed it and pushed a commit to your branch.

Reply:
@mkoppal-px / @lsrinivas-pure / @kshithijiyer-px Can you please confirm if it is necessary to have at least one node with a storage driver installed?

Reply:
@tthurlapati-px  Why can't we use `getPXNamespace()` here?  

@kphalgun-px At the least you would need 3 nodes for any storage driver to run. Some solutions can run with 1 node as well but it's not advised to have one node deployements in production as there is no way to guarantee HA.

Reply:
@kshithijiyer-px  `getPXNamespace` is a `backupHelper` function and cannot be used from `ssh.go`.

And updating the PX namespace `ssh.go` i.e. `s *SSH` uses can only be done via its function as `s.execPodNamespace` is a private variable.

✅

Reply:
Why can't we move the function to `common.go` so that you can use it in `ssh.go` ?

Reply:
In updateDriver 

```
	for _, n := range nodes {
		if !n.IsStorageDriverInstalled {
			continue
		}
		if err := s.TestConnection(n, node.ConnectionOpts{
			Timeout:         1 * time.Minute,
			TimeBeforeRetry: 10 * time.Second,
		}); err != nil {
			return &node.ErrFailedToTestConnection{
				Node:  n,
				Cause: err.Error(),
			}
		}
	}

	return nil
```

Is it expected for the function to not throw any errors if the storage driver is not installed on all nodes?

Reply:
> Why can't we move the function to `common.go` so that you can use it in `ssh.go` ?
@kshithijiyer-px 
The path for `ssh.go` is `drivers/node/ssh/ssh.go`, so would it make sense to import it in `tests/common.go` ?

```go
func (s *SSH) updateDriver() error {
	var err error

	// The namespace of the portworx-service in the current context
	execPodNamespace, err := getExecPodNamespace()
	if err != nil {
		return err
	}
	s.execPodNamespace = execPodNamespace
```

Also, `getPXNamespace()` uses `kube-system` by default AND assumes that portworx is installed in the same namespace on both the source and destination cluster, whereas it may be installed on different namespaces on each cluster. The `getExecPodNamespace` actually queries the cluster to figure out where PXE is installed without assuming it is the same for all clusters.

Reply:
so would it make sense to import `tests/common.go` in it? -> Why wouldn't it make sense? 
What's the issue in calling the same function twice?

Reply:
@kphalgun-px No, ideally it should throw an error.

Reply:
> so would it make sense to import `tests/common.go` in it? -> Why wouldn't it make sense? What's the issue in calling the same function twice?

In my understanding, `tests/*` _uses_ the various tools in `drivers/*`, so the imports should always be such that `tests/*` imports stuff in `drivers/*`. Even though one could import stuff in `tests/*` (like `common.go`) into `drivers/*`, that would basically make `drivers/*` dependent on `tests/*` which doesn't make any **logical** sense to me as drivers are obviously _independent_ of any tests; _even though_ from a programming POV, it's possible (afaik).

Reply:
@mkoppal-px @lsrinivas-pure What would you suggest here? I don't understand the point of having multiple functions to do the same thing.

Reply:
Actually, as far as I can see, `getPXNamespace()` defined in `backup_helper` is only being used 2 times in `backup_resiliency_test`, so we can technically get rid of the function all together and replace it with `ssh.getExecPodNamespace()`. I suggest this because `ssh.getExecPodNamespace()` is a thoroughly implemented version.

The only problem would be if the functions using `getPXNamespace` _speciafically_ want to use the environment variable value of `PX_NAMESPACE` as this is exactly what `getPXNamespace` does. If this is the case, then `getPXNamespace` and `ssh.getExecPodNamespace()` are actually doing different things; the former uses an env var whereas the latter actually determines the namespace at runtime.